[#4] Post-API 테스트코드 완성 및 문서화 추가

gombasan opened the pull request:
# 작업 사항
## Post 테스트코드 작성
- #27 해당 PR 에 이어서 작업한 내용입니다!

[feat : PostControllerTest 추가](https://github.com/Clover-Habiters/backend/commit/957dc000cd2412a86ad47c9a721918a4605d12ee)  …
- PostController 테스트 코드를 추가하였습니다.
- 검색을 할 경우 fulltext 로 DB 에 native 쿼리를 사용함에 따라 트랜잭션을 분리하여 처리하도록 변경하였습니다.
- 그 외의 테스트는 기존 트랜잭션에 합류 하도록 전파 설정을
REQUIRED 로 설정하였습니다.

## 문서화 추가
- #32 해당 PR 에 이어서 작업한 내용입니다!

[feat : Post-API 문서화 추가](https://github.com/Clover-Habiters/backend/commit/0cf6f4d449e4c40517cc72c63792489356880a2f)
- Post API 를 문서화 하기 위한 adoc 파일을 추가하였습니다.
- Post 검색 에서 사용하는 search-type 도 같이 추가하였습니다.

[chore : enum 문서화 업데이트](https://github.com/Clover-Habiters/backend/commit/b84bbf07da7e7678f412b7f7cd2bd0b8e640dde6) 
- Post 검색 에서 사용하는 search-type enum 을 문서화 하기 위하여 searchType 을 추가하였습니다.

[chore : 불필요 헤더 제거 및 함수 패키징 변경](https://github.com/Clover-Habiters/backend/commit/decbad98d7fcf29bee314ad0f47b9732c8a0bf05)  
- http-response 를 전달할 경우 불필요한 헤더정보로 인해 필요한 헤더정보가 가려지는것 같아, 불필요한 헤더는 지우도록 하였습니다.
- 기존 문서화에 있는 static 메서드를 Util 메서드로 이동하였습니다.

[feat : pageable 및 commons 코드 문서화 추가](https://github.com/Clover-Habiters/backend/commit/a7f12f0197985a1830731aa9c38a426e2f2de31d)  
- content 가 비어있는 Page 를 반환하여 문서화를 진행합니다.
- 공통으로 사용하는 `code`,`message`,`data` 를 문서화 하도록 추가하였습니다.
- Enum 보다 공통적인 부분을 문서화 해준다고 생각하여
클래스 명을 `EnumDocumentation` -> `CommonDocumentation` 으로 변경하였습니다.

[chore : 클래스명 변경](https://github.com/Clover-Habiters/backend/commit/123c5dcd7f8aa1ddefb91c5660dc63f8a1915679)  
- `EnumDocs` -> `EnumCode` 로 변경하였습니다.

[chore : 코드 변경에 따른 adoc 문서 수정](https://github.com/Clover-Habiters/backend/commit/7c7a2a1b78252edb6bcf9acc6c75e1324efcddfb)  
- 클래스명 변경에 따른 snippet 경로를 수정하였습니다.
- pageable, commons 내용을 반영하였습니다.
- PostAPI 를 추가하였습니다.
- response 헤더 템플릿 스니펫을 추가하였습니다.
- query-parameters 의 경우 필수값이 아닌경우 기본값을 보여주는 컬럼을 추가하였습니다.

## 특이사항
PostResponse 를 전달할 때, 시,분,초 도 필요하다고 하여 날짜 포맷을 변경하였습니다!
n분전, n초전 까지 표현하기 위함이라구 합니다!

gombasan committed:
feat : Post-API 문서화 추가

- Post API 를 문서화 하기 위한 adoc 파일을 추가하였습니다.
- Post 검색 에서 사용하는 search-type 도 같이 추가하였습니다.

gombasan committed:
chore : enum 문서화 업데이트

- Post 검색 에서 사용하는 search-type enum 을 문서화 하기 위하여 searchType 을 추가하였습니다.

gombasan committed:
feat : PostControllerTest 추가

- PostController 테스트 코드를 추가하였습니다.
- 검색을 할 경우 fulltext 로 DB 에 native 쿼리를 사용함에 따라 트랜잭션을 분리하여 처리하도록 변경하였습니다.
- 그 외의 테스트는 기존 트랜잭션에 합류 하도록 전파 설정을
REQUIRED 로 설정하였습니다.

gombasan committed:
chore : 불필요 헤더 제거 및 함수 패키징 변경

- http-response 를 전달할 경우 불필요한 헤더정보로 인해 필요한 헤더정보가 가려지는것 같아, 불필요한 헤더는 지우도록 하였습니다.
- 기존 문서화에 있는 static 메서드를 Util 메서드로 이동하였습니다.

gombasan committed:
feat : pageable 및 commons 코드 문서화 추가

- content 가 비어있는 Page 를 반환하여 문서화를 진행합니다.
- 공통으로 사용하는 `code`,`message`,`data` 를 문서화 하도록 추가하였습니다.
- Enum 보다 공통적인 부분을 문서화 해준다고 생각하여
클래스 명을 `EnumDocumentation` -> `CommonDocumentation` 으로 변경하였습니다.

gombasan committed:
chore : 클래스명 변경

- `EnumDocs` -> `EnumCode` 로 변경하였습니다.

gombasan committed:
chore : 코드 변경에 따른 adoc 문서 수정

- 클래스명 변경에 따른 snippet 경로를 수정하였습니다.
- pageable, commons 내용을 반영하였습니다.
- PostAPI 를 추가하였습니다.
- response 헤더 템플릿 스니펫을 추가하였습니다.
- query-parameters 의 경우 필수값이 아닌경우 기본값을 보여주는 컬럼을 추가하였습니다.

gombasan committed:
chore : searchTest pageable 내용 제거

- searchTest 에 pageabel 내용을 제거하였습니다.

gombasan committed:
chore : TestPost Overload 함수 추가

- 테스트용 게시글을 생성할때, 타이틀과 컨텐츠를 커스텀 하여 생성할 수 있는 함수를 추가하였습니다.

gombasan committed:
chore : 날짜 포맷 변경

- 프론트측에서 시,분,초 까지의 내용의 필요하다 하여 포맷을 변경하였습니다.

JoosungKwon commented on src/main/java/com/clover/habbittracker/domain/post/dto/PostResponse.java:
```diff
@@ -15,7 +15,7 @@ public record PostResponse(
 	Integer numOfComments,
 	Integer numOfEmojis,
-	@JsonFormat(shape = Shape.STRING, pattern = "yyyy-MM-dd", timezone = "Asia/Seoul")
+	@JsonFormat(shape = Shape.STRING, pattern = "yyyy-MM-dd HH:mm:ss", timezone = "Asia/Seoul")
```
LGTM 👍

JoosungKwon commented on src/main/java/com/clover/habbittracker/domain/post/dto/PostSearchCondition.java:
```diff
@@ -20,8 +21,16 @@ public class PostSearchCondition {
 
-	public enum SearchType {
-		ALL, TITLE, CONTENT;
+	public enum SearchType implements RestDocsEnum {
```
이건 뭔가요..? 뭔가 그렇게 좋아 보이진 않다고 생각하시나요..? 다른 Enum에도 전부 적용되었나요?

gombasan replied:
아! enum 의 설명부분을 뽑아내서 문서화 하기 위해
`getDescription()` ,`getName()` 을 인터페이스로 만들어서 구현하도록 변경했어요!
다른 enum 클래스에도 문서화가 필요한 클래스는 적용시켰습니다..!

JoosungKwon replied:
반드시 해당 방식을 적용 해야하나요? 🤔  테스트 코드와 프러덕트 코드가 분리되면 좋을 것 같다고 생각합니다~ ㅎㅎ

gombasan replied:
```java
private Map<String, String> getDocs(RestDocsEnum[] enumTypes) {
	return Arrays.stream(enumTypes)
		.collect(Collectors.toMap(RestDocsEnum::getName, RestDocsEnum::getDescription));
}
```
이 메서드로 문서화를 위해서 동일 타입, 구현 까지 한꺼번에 처리하고자 했던거라서...
저두 분리하는것이 더 옳다고 생각합니다!

한번 도전해보도록 하겠습니다~!

JoosungKwon commented on src/test/java/com/clover/habbittracker/domain/post/api/PostControllerTest.java:
```diff
@@ -0,0 +1,246 @@
+
+import jakarta.transaction.Transactional;
+
+@Transactional(value = Transactional.TxType.REQUIRED)
```
TxType은 뭔가요..?

gombasan replied:
트랜잭션의 타입으로 알고있습니다! 전파 옵션을 사용하려고 쓰니까.. 
사실 인텔리제이가 추천해줘서 일단 썼어요...😂😂

JoosungKwon replied:
아하 처음 본것 같네요! 보통은 `Transactional.Propagation.REQUIRES_NEW`를 사용한다고 알고 있었습니다~

gombasan replied:
ㅋㅋㅋㅋㅋ 그게 맞아요! 알고보니까 제가 spring 트랜잭션이 아닌 다른 트랜잭션을 사용했더라구요!

[chore : Transactional import 구문 수정](https://github.com/Clover-Habiters/backend/pull/38/commits/8cd0a4bbde09d3e047e90fa599aca8aff426d031)
수정했습니다~!~!~!

JoosungKwon commented on src/test/java/com/clover/habbittracker/domain/post/api/PostControllerTest.java:
```diff
@@ -0,0 +1,246 @@
+		mockMvc.perform(get("/posts")
+				.header("Authorization", "Bearer " + accessToken))
+			.andExpect(status().isOk())
+			.andExpect(jsonPath("$.data[*].id").exists())
```
아.... `[*]` 이렇게 사용하는 거였군요! 👍👍

gombasan replied:
넵! 정규식에서 `[*]` 은 배열에 모든 내용이라구 합니다!

JoosungKwon replied:
몰랐었는데 덕분에 알게됐습니다~!! 저도 적용해볼께요~~ 😄

JoosungKwon commented on src/test/java/com/clover/habbittracker/global/restdocs/common/CommonDocumentation.java:
```diff
@@ -0,0 +1,117 @@
+import com.fasterxml.jackson.core.type.TypeReference;
+
+@DisplayName("RestDocs 공통 코드 문서화")
+public class CommonDocumentation extends RestDocsSupport {
```
크... 고생 많으셨습니다~!!!!!

gombasan replied:
잘만든건지 계속 의문이 들지만 감사합니다 ㅎㅎㅎㅎ

JoosungKwon replied:
아니다 싶으면 리팩터링을 하면 되죠~ 원래 조금씩 개선해 나가는 겁니다!

JoosungKwon commented on src/test/java/com/clover/habbittracker/global/restdocs/util/ApiDocumentUtils.java:
```diff
@@ -16,6 +22,32 @@ public static OperationRequestPreprocessor getDocumentRequest() {
+			),
+			prettyPrint());
+	}
+
+	// 커스텀 템플릿 사용을 위한 함수
+	public static CustomResponseFieldsSnippet customResponseFields
+	(String type,
+		PayloadSubsectionExtractor<?> subsectionExtractor,
+		Map<String, Object> attributes, FieldDescriptor... descriptors) {
+		return new CustomResponseFieldsSnippet(type, subsectionExtractor, Arrays.asList(descriptors), attributes
+			, true);
+	}
+
+	// Map으로 넘어온 enumValue를 fieldWithPath로 변경하여 리턴
+	public static FieldDescriptor[] enumConvertFieldDescriptor(Map<String, String> enumValues) {
+		return enumValues.entrySet().stream()
+			.map(x -> fieldWithPath(x.getKey()).description(x.getValue()))
+			.toArray(FieldDescriptor[]::new);
 	}
```
이런건 어디 자료에서 볼 수 있나요?

gombasan replied:
해당 내용은 
https://backtony.github.io/spring/2021-10-15-spring-test-3/#%EA%B3%B5%ED%86%B5-%EC%BD%94%EB%93%9C-%EB%AC%B8%EC%84%9C%ED%99%94

위에 블로그 글을 참고해서 작성했어요...!
거의 복붙하듯이 따라하긴 했습니다....😢😢

JoosungKwon replied:
와 이분 엄청 잘 정리했네요.. 저도 참고 하겠습니다! 조금씩 알아가면 되는거죠..😄

JoosungKwon commented on src/test/java/com/clover/habbittracker/util/CustomTransaction.java:
```diff
@@ -0,0 +1,11 @@
+
+import org.springframework.transaction.TransactionDefinition;
+
+public class CustomTransaction implements TransactionDefinition {
+
+	@Override
+	public int getPropagationBehavior() {
+		return TransactionDefinition.PROPAGATION_REQUIRES_NEW;
+	}
+}
```
굿 👍 저도 잘 참고해보고 따라 해보겠습니다~!!

JoosungKwon reviewed (approved):
테스트 고수가 되신 것 같네요..☺️ 저두 한수 배워갑니다~!!! 고생하셨습니다!

gombasan commented:
> 테스트 고수가 되신 것 같네요..☺️ 저두 한수 배워갑니다~!!! 고생하셨습니다!

감사합니다~~!! ㅎㅎㅎ
더 갈고 닦도록 하겠습니다 ㅎㅎ

JoosungKwon commented:
@gombasan 
`Post` 검색 관련 `DDL Schema` 에서 `Full Text Search Index`의 알고리즘을 명시하는 부분이 빠진 것 같아서 추가했습니다~ 

[`chore: schema.sql에 전문 검색 알고리즘 명시`](https://github.com/Clover-Habiters/backend/pull/38/commits/982a3323cc0820670cb91b5c78c765e2ba6b117f)

JoosungKwon committed:
chore: schema.sql에 전문 검색 알고리즘 명시

gombasan commented:
> @gombasan `Post` 검색 관련 `DDL Schema` 에서 `Full Text Search Index`의 알고리즘을 명시하는 부분이 빠진 것 같아서 추가했습니다~
> 
> [`chore: schema.sql에 전문 검색 알고리즘 명시`](https://github.com/Clover-Habiters/backend/pull/38/commits/982a3323cc0820670cb91b5c78c765e2ba6b117f)

@JoosungKwon 
오호 수정 감사합니다~~! ㅎㅎ
몇가지 수정하고 오늘 저녁에 머지 하도록 하겠습니다!

gombasan committed:
chore : Transactional import 구문 수정

- Jakarta EE의 트랜잭션 import를 Spring Framework의 Propagation을 이용한 구문으로 변경하였습니다.

gombasan committed:
refactor : 게시글 검색 테스트 코드 수정

- 기존 검색 테스트를 전부 customTransaction 을 활용하여 테스트 하도록 수정하였습니다.
- 기존 beforeTransaction 은 트랜잭션이 시작하기전 실행되므로, 멤버를 생성하기 전 게시글을 먼저 생성하는 문제가 있었습니다.
- 트랜잭션을 분리하는 방법으로 해결하였습니다.

gombasan committed:
chore : 게시글 반환 시 memberId 추가

- 게시글 작성자를 확인하기 위하여 memberId 를 추가하였습니다.

gombasan commented:
@JoosungKwon 
추가로 작업 내역들이 있습니다!

[refactor : 게시글 검색 테스트 코드 수정](https://github.com/Clover-Habiters/backend/pull/38/commits/b0f2d9edf81b54652ba88621eb0495cf372e9945) 

- 기존 검색 테스트를 전부 customTransaction 을 활용하여 테스트 하도록 수정하였습니다.
- 기존 beforeTransaction 은 트랜잭션이 시작하기전 실행되므로, 멤버를 생성하기 전 게시글을 먼저 생성하는 문제가 있었습니다.
- 트랜잭션을 분리하는 방법으로 해결하였습니다.

[chore : 게시글 반환 시 memberId 추가](https://github.com/Clover-Habiters/backend/pull/38/commits/8d2fb8fbe32f009919709ea028130ef7ebde9868) 

- 게시글 작성자를 확인하기 위하여 memberId 를 추가하였습니다.
